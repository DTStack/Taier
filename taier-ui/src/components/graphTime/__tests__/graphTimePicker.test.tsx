import { cleanup, fireEvent, render } from '@testing-library/react';
import moment from 'moment';
import GraphTimePicker from '../graphTimePicker';

function openPicker(container: HTMLElement, index = 0) {
    const input = container.querySelectorAll('input')[index];
    fireEvent.mouseDown(input);
    fireEvent.focus(input);
}

function findCell(text: string | number, index = 0) {
    const table = document.querySelectorAll('table')[index];

    const array = Array.from(table.querySelectorAll('td'));
    for (let j = 0; j < array.length; j += 1) {
        const td = array[j];
        if (td.textContent === String(text) && td.className.includes('-in-view')) {
            return td;
        }
    }
}

function selectCell(text: string | number, index = 0) {
    const td = findCell(text, index);
    fireEvent.click(td!);

    return td;
}

function confirmOK() {
    fireEvent.click(document.querySelector('.ant-picker-ok > *')!);
}

describe('Test GraphTimePicker Component', () => {
    beforeEach(() => {
        cleanup();
        document.body.innerHTML = '';
    });

    it('Should match snapshot', () => {
        const { asFragment } = render(<GraphTimePicker timeRange="10m" value={moment(new Date('2022-01-20'))} />);
        expect(asFragment()).toMatchSnapshot();
    });

    it('Should call onChange event handler', async () => {
        const fn = jest.fn();
        const { container } = render(
            <GraphTimePicker timeRange="10m" value={moment(new Date('2022-01-20'))} onChange={fn} />
        );

        expect(fn).not.toBeCalled();
        openPicker(container);
        selectCell(3);
        confirmOK();

        // Once for onChange, another for onChange
        expect(fn).toBeCalledTimes(2);
        expect((fn.mock.calls[0][0] as moment.Moment).isSame(moment(new Date('2022-01-03')))).toBeTruthy();
    });

    it('Should disabled the day after today and time', () => {
        const original = Date.now;
        Date.now = jest.fn(() => new Date('2022-01-21T12:33:37.000').valueOf());

        const { container } = render(
            <GraphTimePicker timeRange="10m" value={moment(new Date('2022-01-21T12:33:37.000'))} />
        );

        openPicker(container);

        const dates = Array.from(document.querySelector('table')?.querySelectorAll('td') || []);
        const idx = dates.findIndex((d) => d.textContent === '21');

        // Expect the days after today(today is 2022-01-21) are all disabled
        expect(dates.slice(idx + 1).every((d) => d.classList.contains('ant-picker-cell-disabled'))).toBeTruthy();

        const hours = Array.from(
            document.querySelectorAll('ul.ant-picker-time-panel-column')[0]?.querySelectorAll('li') || []
        );
        // Expect the hours after now are all disabled
        expect(hours.slice(13).every((h) => h.classList.contains('ant-picker-time-panel-cell-disabled'))).toBeTruthy();

        const minutes = Array.from(
            document.querySelectorAll('ul.ant-picker-time-panel-column')[1]?.querySelectorAll('li') || []
        );
        expect(
            minutes.slice(34).every((m) => m.classList.contains('ant-picker-time-panel-cell-disabled'))
        ).toBeTruthy();

        const seconds = Array.from(
            document.querySelectorAll('ul.ant-picker-time-panel-column')[2]?.querySelectorAll('li') || []
        );
        expect(
            seconds.slice(38).every((s) => s.classList.contains('ant-picker-time-panel-cell-disabled'))
        ).toBeTruthy();

        Date.now = original;
    });

    it('Should support to change value via button', () => {
        const fn = jest.fn();
        const { container } = render(
            <GraphTimePicker timeRange="10m" value={moment(new Date('2022-01-21T12:33:37.000'))} onChange={fn} />
        );

        fireEvent.click(container.querySelector('.btn-prev')!);
        expect(fn).toBeCalledTimes(1);
        expect((fn.mock.calls[0][0] as moment.Moment).format('YYYY-MM-DD hh:mm:ss')).toBe('2022-01-21 12:23:37');

        fireEvent.click(container.querySelector('.btn-next')!);
        expect(fn).toBeCalledTimes(2);
        expect((fn.mock.calls[1][0] as moment.Moment).format('YYYY-MM-DD hh:mm:ss')).toBe('2022-01-21 12:43:37');
    });
});
