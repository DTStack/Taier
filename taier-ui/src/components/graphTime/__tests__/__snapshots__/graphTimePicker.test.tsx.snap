// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test GraphTimePicker Component Should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="graph-time graph-time-picker"
  >
    <button
      class="ant-btn ant-btn-default btn-prev"
      type="button"
    >
      <span
        aria-label="left"
        class="anticon anticon-left"
        role="img"
      >
        <svg
          aria-hidden="true"
          data-icon="left"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M724 218.3V141c0-6.7-7.7-10.4-12.9-6.3L260.3 486.8a31.86 31.86 0 000 50.3l450.8 352.1c5.3 4.1 12.9.4 12.9-6.3v-77.3c0-4.9-2.3-9.6-6.1-12.6l-360-281 360-281.1c3.8-3 6.1-7.7 6.1-12.6z"
          />
        </svg>
      </span>
    </button>
    <div
      class="ant-picker"
    >
      <div
        class="ant-picker-input"
      >
        <input
          autocomplete="off"
          placeholder="End Time"
          readonly=""
          size="21"
          title="2022-01-20 00:00:00"
          value="2022-01-20 00:00:00"
        />
        <span
          class="ant-picker-suffix"
        >
          <span
            aria-label="calendar"
            class="anticon anticon-calendar"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="calendar"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M880 184H712v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H384v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H144c-17.7 0-32 14.3-32 32v664c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V216c0-17.7-14.3-32-32-32zm-40 656H184V460h656v380zM184 392V256h128v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h256v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h128v136H184z"
              />
            </svg>
          </span>
        </span>
        <span
          class="ant-picker-clear"
          role="button"
        >
          <span
            aria-label="close-circle"
            class="anticon anticon-close-circle"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="close-circle"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
              />
            </svg>
          </span>
        </span>
      </div>
    </div>
    <button
      class="ant-btn ant-btn-default btn-next"
      type="button"
    >
      <span
        aria-label="right"
        class="anticon anticon-right"
        role="img"
      >
        <svg
          aria-hidden="true"
          data-icon="right"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
          />
        </svg>
      </span>
    </button>
  </div>
</DocumentFragment>
`;
